src/write.rs

```diff
@@ -548,9 +611,61 @@ where
                     }
                 }
             }
+            Operation::Retain(ref key, ref predicate) => {
+                if let Some(e) = inner.data.get_mut(key) {
+                    let mut del = 0;
+
+                    let len = e.len();
+
+                    for i in 0..len {
+                        if !predicate.eval(unsafe { e.get_unchecked(i) }) {
+                            del += 1;
+                        } else {
+                            e.swap(i - del, i);
+                        }
+                    }
```

Comment:
This could use a comment describing what's going on. You're basically moving all the things we're dropping to the end, and then forgetting them, right? Writing that out would be helpful.

Reply:
Yeah, it's "bubbling up" the elements we wish to remove, then popping them all off at the end. This was almost directly copied from `smallvec`'s `retain` implementation.

Reply:
I think it would be helpful to add a comment + maybe even linking to `Vec::retain` and then say that we "do pretty much the same", but have to do it manually so that we don't drop anything

Reply:
I think linking to `Vec::retain` and saying in the same place that we can't use `.retain` because of dropping would be a good change. Makes the code easier to read for someone who's not familiar with it!